src/serde/ser.rs

```diff
@@ -0,0 +1,621 @@
+use std::io::Cursor;
+
+use crate::ion_writer::IonWriter;
+use crate::result::IonFailure;
+use crate::serde::decimal::TUNNELED_DECIMAL_TYPE_NAME;
+use crate::serde::timestamp::TUNNELED_TIMESTAMP_TYPE_NAME;
+use crate::types::Int;
+use crate::{
+    BinaryWriterBuilder, Decimal, IonError, IonResult, IonType, TextKind, TextWriterBuilder,
+    Timestamp,
+};
+use serde::ser::Impossible;
+use serde::{ser, Serialize};
+
+/// Serialize an object into pretty formatted Ion text
+pub fn to_pretty<T>(value: &T) -> IonResult<String>
+where
+    T: Serialize,
+{
+    let mut cursor = Cursor::new(Vec::new());
+    let mut serializer = Serializer {
+        writer: TextWriterBuilder::pretty().build(&mut cursor)?,
+    };
+
+    value.serialize(&mut serializer)?;
+    serializer.writer.flush()?;
+    drop(serializer);
+
+    let bytes = cursor.get_ref().clone();
+
+    match String::from_utf8(bytes) {
+        Ok(data) => Ok(data),
+        Err(e) => IonResult::encoding_error(e.to_string()),
+    }
+}
+
+/// Serialize an object into compact Ion text format
+pub fn to_string<T>(value: &T) -> IonResult<String>
+where
+    T: Serialize,
+{
+    let mut cursor = Cursor::new(Vec::new());
+    let mut serializer = Serializer {
+        writer: TextWriterBuilder::new(TextKind::Compact).build(&mut cursor)?,
+    };
+
+    value.serialize(&mut serializer)?;
+    serializer.writer.flush()?;
+    drop(serializer);
+
+    let bytes = cursor.get_ref().clone();
+
+    match String::from_utf8(bytes) {
+        Ok(data) => Ok(data),
+        Err(e) => IonResult::encoding_error(e.to_string()),
+    }
+}
+
+/// Serialize an object into Ion binary format
+pub fn to_binary<T>(value: &T) -> IonResult<Vec<u8>>
+where
+    T: Serialize,
+{
+    let mut cursor = Cursor::new(Vec::new());
+    let mut serializer = Serializer {
+        writer: BinaryWriterBuilder::new().build(&mut cursor)?,
+    };
+
+    value.serialize(&mut serializer)?;
+    serializer.writer.flush()?;
+    drop(serializer);
+
+    Ok(cursor.get_ref().clone())
+}
+
+/// Implements a standard serializer for Ion
+pub struct Serializer<E> {
+    pub(crate) writer: E,
+}
```

Comment:
Having a way to pass in a writer of my own would be helpful.

Reply:
https://github.com/amazon-ion/ion-rust/issues/646